docs/src/lib/indexing.md

```diff
@@ -76,6 +76,124 @@ For performance reasons, accessing, via `getindex` or `view`, a single `row` and
+      which case it is enough that number of rows in each entry of `v` is constant;
+    * number of columns in `v` must be equal to length of `cols`
+    * before making any assignment `cols` is transformed to `Symbols` using `names(df)`
+      (which means that passing `Integer` or `Bool` will fail for missing columns,
+      but adding new columns as `Symbol` is allowed)
+    * then we perform `df[col] = v[col]` for `col in cols`
+* if `v` is an `AbstractMatrix` the same process is performed but we perform
+  `df[col] = v[:, i]` for `(i, col) in enumerate(col)` instead
+
+> `df[cols] .= v`
+
+* before making any assignment `cols` is transformed to `Symbols` using `names(df)`
+  (which means that passing `Integer` or `Bool` will fail for missing columns,
+  but adding new columns as `Symbol` is allowed)
+* if `v` is a `DataFrame` or `NamedTuple` or `AbstractDict`:
+    * number of columns in `v` must be equal to length of `cols`
+    * before making any assignment `cols` is transformed to `Symbols` using `names(df)`
+      (which means that passing `Integer` or `Bool` will fail for missing columns,
+      but adding new columns as `Symbol` is allowed)
+    * then we perform `df[col] .= v[col]` for `col in cols`
+* if `v` is 0- or 1- dimensional then it is broadcasted to call `df[col] .= v` for all `col in cols`
+* if `v` is 2-dimensional then its dimensions of length 1 are expanded if needed and we perform
+  `df[col] .= v[:, i]` for `(i, col) in enumerate(col)`
+
+> `df[row, col] = v`
+
+* translates to `df[col][row] = v`
+
+> `df[row, col] .= v`
+
+* translates to `df[col][row] .= v`
+
+> `df[row, cols] = v`
+
+* if `v` is a `DataFrameRow` or `NamedTuple` or `AbstractDict` then
+    * number of columns in `v` must be equal to length(cols)
+    * column names in v must be the same as selected by `cols`
+    * an operation `df[row, col] = v[col]` for `col in cols` is performed
+* if `v` is a vector or a tuple:
+    * number of elements in `v` must be equal to length(cols)
+    * an operation `df[row, col] = v[i]` for `(i, col) in enumerate(cols)` is performed
+* otherwise an error is thrown
+
+> `df[row, cols] .= v`
+
+* if `v` is a `DataFrameRow` or `NamedTuple` or `AbstractDict` then
+    * number of columns in `v` must be equal to length(cols)
+    * column names in v must be the same as selected by `cols`
+    * an operation `df[row, col] .= v[col]` for `col in cols` is performed
+* if `v` is 0-dimensional:
+    * an operation `df[row, col] .= v[]` for `col in cols` is performed
+* if `v` is 1-dimensional:
+    * number of elements in `v` must be equal to length(cols)
+    * an operation `df[row, col] .= v[i]` for `(i, col) in enumerate(cols)` is performed
+* otherwise an error is thrown
+
+> `df[rows, col] = v`
+
+* The same rules as for `df[col] = v` but on selected rows and always with copying.
+* Empty data frames are not allowed and no column adding is possible.
```

Comment:
We have `df[col] = v` and `df[col] .= v` for this case. It would render `df[:, col] = v` inconsistent.

And here I exactly try to fix the inconsistency in the current design of DataFrames.jl. See the example how `x` and `y` are treated differently:
```
julia> x = [1,2,3]
3-element Array{Int64,1}:
 1
 2
 3

julia> y = [1,2,3]
3-element Array{Int64,1}:
 1
 2
 3

julia> df = DataFrame(x=x,y=y)
3×2 DataFrame
│ Row │ x     │ y     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 1     │
│ 2   │ 2     │ 2     │
│ 3   │ 3     │ 3     │

julia> df[:, :x] = [0,0,0]
3-element Array{Int64,1}:
 0
 0
 0

julia> df[1:3, :y] = [0,0,0]
3-element Array{Int64,1}:
 0
 0
 0

julia> df
3×2 DataFrame
│ Row │ x     │ y     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 0     │ 0     │
│ 2   │ 0     │ 0     │
│ 3   │ 0     │ 0     │

julia> x
3-element Array{Int64,1}:
 1
 2
 3

julia> y
3-element Array{Int64,1}:
 0
 0
 0

julia> df[:, :x] = 'a':'c'
'a':1:'c'

julia> df
3×2 DataFrame
│ Row │ x    │ y     │
│     │ Char │ Int64 │
├─────┼──────┼───────┤
│ 1   │ 'a'  │ 0     │
│ 2   │ 'b'  │ 0     │
│ 3   │ 'c'  │ 0     │

julia> df[1:3, :y] = 'a':'c'
'a':1:'c'

julia> df
3×2 DataFrame
│ Row │ x    │ y     │
│     │ Char │ Int64 │
├─────┼──────┼───────┤
│ 1   │ 'a'  │ 97    │
│ 2   │ 'b'  │ 98    │
│ 3   │ 'c'  │ 99    │
```

Reply:
I agree this is problematic, but I don't see the relationship with empty data frames (it's been a long time....).

Reply:
The proposed rule is that if we use `df[:, col]` or `df[rows, col]` in general we write to an existing vector in the `DataFrame` in-place. So - naturally it has to fail in an empty `DataFrame`.

Or - from another angle - the proposed rule is that on LHS `df[:, col]` is rewritten as `df[axes(df, 1), col]`.

This approach is consistent with Base, where the manual states [here](https://docs.julialang.org/en/v1/manual/arrays/index.html#man-array-indexing-1):

> This includes `Colon` (`:`) to select all indices within the entire dimension

I would be OK to bend the rules in this case, if there were no other convenient way to assign a new column to a `DataFrame`, but there is one and it is `df[col] = v`.

But if you really think it is useful we can make an exception for an empty `DataFrame`.

Reply:
But if we have e.g.
```julia
df = DataFrame(x=[])
df[:, 1] = []
```
Then what's the problem? We've set all entries in `df.x` to the values from the RHS -- it's just that there aren't any of them. :-)

Reply:
There is no problem 😄. Your `df` is not empty, it has 0-rows, but it is a different story. An empty data frame is returned by the call `DataFrame()`.

Reply:
Note that this distinction of emptyness is very relevant in other cases.

This should work:
```
df = DataFrame()
df.a = [1,2,3]
``
while this should fail
```
df = DataFrame(x=[])
df.a = [1,2,3]
```

An interesting corner case is:
```
julia> df = DataFrame(x=[1,2])
2×1 DataFrame
│ Row │ x     │
│     │ Int64 │
├─────┼───────┤
│ 1   │ 1     │
│ 2   │ 2     │

julia> df.x = [1,2,3,4]
ERROR: ArgumentError: New columns must have the same length as old columns
```
which fails now and I thing it should still fail.

The short rule - an empty data frame `DataFrame()` has an undefined number of rows (not `0`) while any non-empty data frame has a concrete number of rows.

Actually we could consider to make `nrow(DataFrame())` return `nothing` instead of `0`, this would be technically stricter, but I always thought it would not be very useful, so I have not proposed to add this change.

Reply:
OK, then maybe say "with zero columns" to avoid any possible ambiguity.

> Actually we could consider to make `nrow(DataFrame())` return `nothing` instead of `0`, this would be technically stricter, but I always thought it would not be very useful, so I have not proposed to add this change.

Or even `missing`, since it's really unknown. ;-) But `0` is probably the most useful answer.

Reply:
OK - I will replace empty with zero columns everywhere to disambiguate.